examples/rlhf/train_rlhf.py

```diff
@@ -0,0 +1,339 @@
+        resolve_name_or_path(transformer_name_or_path), dropout, device, compile_
+    )
+    ref_model = deepcopy(model).to("cuda:1")
+    ref_model.requires_grad_(False)
+    layers = model.transformer.h
+    num_layers = len(layers)
+    num_unfrozen = int(0.3 * num_layers)
+    for layer in layers[:-num_unfrozen]:
+        layer.requires_grad_(False)
+
+    reward_model = init_reward_model(
+        reward_model_path=resolve_name_or_path(reward_model_cfg.name_or_path),
+        device=device,
+        compile_=compile_,
+    )
+    reward_model.eval()
+    reward_model.requires_grad_(False)
+
+    adv_fn = GAE(
+        value_network=critic, gamma=0.99, lmbda=0.95, average_gae=True, shifted=True
+    )
+    loss_fn = ClipPPOLoss(actor, critic_head)
+
+    test_prompt = next(val_loader)
+    estimate_reward = create_reward_estimator(
+        eval_iters,
+        episode_length,
+        reward_model,
+        test_prompt,
+        ctx,
+        logger=query_logger,
+        ref_model=ref_model,
+    )
+
+    optimizer = torch.optim.AdamW(
+        [p for p in loss_fn.parameters() if p.requires_grad], **train_cfg.optimizer
+    )
+    scheduler = None
+    if train_cfg.decay_lr:
+        scheduler = CosineAnnealingLR(optimizer, **train_cfg.scheduler)
+
+    rb = TensorDictReplayBuffer(
+        storage=LazyTensorStorage(episode_length * num_rollouts_per_epoch),
+        batch_size=episode_length * batch_size,
+        sampler=SamplerWithoutReplacement(),
+        prefetch=10,
+    )
+    rb_ppo = TensorDictReplayBuffer(
+        storage=LazyTensorStorage(episode_length * batch_size),
+        batch_size=ppo_batch_size,
+        sampler=SamplerWithoutReplacement(),
+        prefetch=10,
+    )
+
+    rollout_from_model = RolloutFromModel(model, ref_model, reward_model)
+
+    best_val_reward = float("-inf")
+    it = 0  # it is equivalent to batch_size number of episodes
+    with tqdm(total=int(max_epochs * num_rollouts_per_epoch / batch_size)) as pbar:
+        for _epoch in range(1, max_epochs + 1):
+            rb.empty()
+            rollout_rewards = []
+            rollout_kl = []
+            kl_controller = AdaptiveKLController(0.1, 6, 10000)
+            for _ in range(0, num_rollouts_per_epoch, batch_size):
+                batch = next(train_loader)
+                td = rollout_from_model.rollout_from_data(
+                    batch, kl_coef=kl_controller.value
+                )
+                with torch.no_grad(), ctx:
+                    # moving this to within epoch
+                    adv_fn(td)
+                # it's possible we didn't fill the replay buffer in the last iteration if
+                # generation stopped early, so we empty first before repopulating
+                rb.extend(flatten_td(td))
+                done = td.get(("next", "done"))
+                next_reward = td.get(("next", "reward_raw"))[done]
+                next_kl = td.get(("next", "reward_kl"))[done]
```

Comment:
this can be done more efficiently via
```suggestion
                td_done = td[done.view(td.shape)]
                next_reward = td_done.get(("next", "reward_raw"))
                next_kl = td_done.get(("next", "reward_kl"))
```